### PR TITLE
feat: Add MultiStore SMT manager

### PR DESCRIPTION
**Pull request by h5law**
## Description

This PR introduces the new `MultiStore` SMT manager. This enables the management of numerous SMTs under a single overarching tree.

The `MultiStore` acts as a wrapper around an SMT that includes the name and root hash of the trees it manages. The `Store` interface is a wrapper around a single SMT which tracks the `MultiStore` it is a part of and overwrites the SMT's `Commit` method to not only commit the `Store`'s changes to the database but also update the `MultiStore` with the name and root hash of the `Store`'s tree.

This is useful in areas where a single tree / root hash is needed but different SMT's are desired: for example in a blockchain setting this could replace the root tree and each of the trees it manages could track the transactions, accounts etc. It would also be useful for an implementation of the IBC specification where the `MultiStore` would represent the IBC store as a whole with each `Store` representing a single ICS component such as the ICS-02 store or the ICS-03 store.

### Summary generated by Reviewpad on 31 Jul 23 12:52 UTC
This pull request introduces significant changes to several files related to the `MultiStore` functionality. Here is a summary of the changes:

1. `multi_test.go`: The file now includes test functions and helper functions for testing the `MultiStore` functionality. The test functions cover various operations such as adding a store, inserting a pre-existing store, getting a store, removing a store, performing store operations, committing changes, and generating proofs. Additionally, a helper function `customStoreCreator` is added.

2. `multi.go`: This file implements the `MultiStore` interface and its associated functions. The `multi` struct represents a collection of multiple stores and implements methods such as `AddStore`, `InsertStore`, `GetStore`, `RemoveStore`, `Commit`, `Root`, `Prove`, and `Spec`. The `store` struct represents an individual store within the `MultiStore`, maintaining a reference to the parent `MultiStore` and implementing a `Commit` method that updates the multi-store with the latest root hash.

3. `MultiStore.md`: The Markdown file provides an overview of the `MultiStore` interface and its functionalities. It explains how the `MultiStore` manages multiple stores and describes the operations available for managing stores and trees. It also introduces the `Store` type, which is a wrapper around an `SMT` and includes a `Commit` method to update the multi-store with the root hash of the store. Additionally, it discusses the `TreeSpec` type, which specifies the hashing functions and maximum depth of the tree.

These changes significantly enhance the functionality and usage of the `MultiStore` concept, allowing for better management and manipulation of multiple stores within the system.

## Type of change

Please mark the relevant option(s):

- [x] New feature, functionality or library
- [ ] Bug fix
- [ ] Code health or cleanup
- [ ] Major breaking change
- [ ] Documentation
- [ ] Other 

## List of changes

- Add the `MultiStore` and `Store` interfaces
- Add the `multi` and `store` structs to implement the interfaces
- Add unit tests
- Add documentation

## Testing

- [x] **Task specific tests or benchmarks**: `go test ...`
- [x] **New tests or benchmarks**: `go test ...`
- [x] **All tests**: `go test -v`

## Required Checklist

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added, or updated, [`godoc` format comments](https://go.dev/blog/godoc) on touched members (see: [tip.golang.org/doc/comment](https://tip.golang.org/doc/comment))
- [x] I have tested my changes using the available tooling

### If Applicable Checklist

- [x] Update any relevant README(s)
- [x] Add or update any relevant or supporting [mermaid](https://mermaid-js.github.io/mermaid/) diagrams
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: multi.go [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+// Spec returns the TreeSpec for the MultiStore tree
+func (m *multi) Spec() *TreeSpec {
+	return m.tree.Spec()
+}
+
+type store struct {
+	*SMT
+	name      string
+	nodeStore MapStore
+	multi     *multi
```
Comment (h5law) [inline multi.go:109]: Should this be an interface?

File: multi.go [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+	_ Store      = (*store)(nil)
+
+	// StoreCreator is a function that creates a new Store for the given MultiStore
+	StoreCreator = func(name string, ms MultiStore) (Store, MapStore) {
+		db := NewSimpleMap()
+		return NewStore(name, ms, db, sha256.New()), db
+	}
+)
+
+type multi struct {
```
Comment (h5law) [inline multi.go:20]: Should the fields be exported and have getters so you can plug and play your own and pass it as an interface into the store?

**Review by Olshansk — CHANGES_REQUESTED** [review]
@h5law Before I review this, can you update the description with the motivation for this?

When I saw this:
> The `MultiStore` interface grants the easy management of numerous different SMTs.

It made me feel like the multistore is outside the scope of the `smt` repo and is probably something a user of this repo would manage.

**Comment by h5law** [comment]
Closing this as the MultiStore is not longer a priority.

For future reference this should be rescoped to allow the SMT to preform silimar functions to the IAVL+ tree in the context of the `CommitMultiStore` etc used in the `cosmos-sdk` [1]. Enabling the correct features for this to be plugged in place of the IAVL+ tree will be a win.

[1] https://docs.cosmos.network/v0.45/core/store.html#commitmultistore

**Pull request closed without merging**